### ISSUE TITLE
Add examples in the documentation of Gather

### ISSUE THREAD

**Issue by fengyuentau** [onnx/onnx#5619]
# Bug Report

### Is the issue related to model conversion?

No.

### Describe the bug

In current Gather documentation, there are two examples, one of which has the following expression:

> If `axis = 1`, let `k = indices[i_{0}, ..., i_{q-1}]` then `output[j_{0}, i_{0}, ..., i_{q-1}, j_{1}, ..., j_{r-2}] = input[j_{0}, k, j_{1}, ..., j_{r-2}]`

It is incorrect for the given example below:

```python
import numpy as np

data = np.random.randn(1, 3, 4).astype(np.float32)
indices = np.array([[0, 2]])
y = np.take(data, indices, axis=1)

print(data.shape)
print(indices)
print(y.shape)
```

output:

```bash
(1, 3, 4)
[[0 2]]
(1, 1, 2, 4)
```

Obviously, the dimension `3` from `(1, 3, 4)` is sort of extended (or replaced?) by the shape of `indices` which is `[1, 2]`, and the shape of output is `(1, 1, 2, 4)`. However, according to the quoted expression, it should be of shape `j_{0}, i_{0}, ..., i_{q-1}, j_{1}, ..., j_{r-2}`, which is `(1, 1, 2, 3)` instead (noted that r=3, so r-2=3-2=1). The expression does not give the same result as the code.

The expressions should be corrected as follows:

- If `axis = 0`, let `k = indices[i_{0}, ..., i_{q-1}]`
then `output[i_{0}, ..., i_{q-1}, j_{axis+1}, ..., j_{r-1}] = input[k , j_{axis+1}, ..., j_{r-1}]`
- If `axis = 1`, let `k = indices[i_{0}, ..., i_{q-1}]` then `output[j_{0}, i_{0}, ..., i_{q-1}, j_{axis+1}, ..., j_{r-1}] = input[j_{0}, k, j_{axis+1}, ..., j_{r-1}]`
- A more general one is `output[j_{0}, ..., j_{axis-1}, i_{0}, ..., i_{q-1}, j_{axis+1}, ..., j_{r-1}] = input[j_{0}, ..., j_{axis-1}, k, j_{axis+1}, ..., j_{r-1}]`, if `axis-1<0` then the part `j_{0}, ..., j_{axis-1}` does not exists, if `axis+1>r-1` then the part `j_{axis+1}, ..., j_{r-1}` does not exist.

@justinchuby Could you verify? If the problem is true, I would like to submit a pull request to correct the expression.

**Comment by justinchuby** [comment]
cc @gramalingam @xadupre @jcwchen

**Comment by justinchuby** [comment]
Preliminary test with ONNX Runtime:

```python
In [1]: from onnxscript import opset18 as op

In [2]: import numpy as np
   ...: 
   ...: data = np.random.randn(1, 3, 4).astype(np.float32)
   ...: indices = np.array([[0, 2]])
   ...: y = np.take(data, indices, axis=1)
   ...: 
   ...: print(data.shape)
   ...: print(indices)
   ...: print(y.shape)
   ...: 
(1, 3, 4)
[[0 2]]
(1, 1, 2, 4)

In [3]: op.Gather(data, indices)
2023-09-23 06:46:54.961557 [E:onnxruntime:, sequential_executor.cc:514 ExecuteKernel] Non-zero status code returned while running Gather node. Name:'' Status Message: indices element out of data bounds, idx=2 must be within the inclusive range [-1,0]
---------------------------------------------------------------------------
InvalidArgument                           Traceback (most recent call last)
Cell In[3], line 1
----> 1 op.Gather(data, indices)

File ~/Documents/GitHub/onnx/venv/lib/python3.11/site-packages/onnxscript/onnx_opset/_impl/opset13.py:941, in Opset13.Gather(self, data, indices, axis)
    939 schema = get_schema("Gather", 13, "")
    940 op = Op(self, "Gather", schema)
--> 941 return op(*self._prepare_inputs(schema, data, indices), axis=axis)

File ~/Documents/GitHub/onnx/venv/lib/python3.11/site-packages/onnxscript/values.py:297, in Op.__call__(self, *args, **kwargs)
    293 if schema is None:
    294     raise RuntimeError(
    295         f"Op '{self.name}' does not have an OpSchema and cannot be evaluated."
    296     )
--> 297 return evaluator.default().eval(schema, args, kwargs)

File ~/Documents/GitHub/onnx/venv/lib/python3.11/site-packages/onnxscript/evaluator.py:196, in BaseEvaluator.eval(self, schema, inputs, attributes)
    194 attributes, closure = self.adapt_attributes(schema, attributes)
    195 inputs = self.adapt_inputs(schema, inputs)
--> 196 outputs = self._eval(schema, inputs, attributes, closure)
    197 return self.adapt_outputs(schema, outputs)

File ~/Documents/GitHub/onnx/venv/lib/python3.11/site-packages/onnxscript/evaluator.py:510, in ORTEvaluator._eval(self, schema, inputs, attributes, closure)
    509 def _eval(self, schema, inputs, attributes, closure):
--> 510     return _call_ort(schema, inputs, attributes, closure)

File ~/Documents/GitHub/onnx/venv/lib/python3.11/site-packages/onnxscript/evaluator.py:487, in _call_ort(schema, args, kwargs, implicit_args)
    480     raise EagerModeError(
    481         f"Unable to create onnxruntime InferenceSession "
    482         f"for executing {schema.domain}.{schema.name} op "
    483         f"with onnx model\n{utils.proto2text(model)}"
    484     ) from e
    486 try:
--> 487     result = session.run(None, session_run_input)
    488 except (RuntimeError, Fail) as e:
    489     raise EagerModeError(
    490         f"Unable to execute model operator {schema.name!r} due to {e!r}"
    491         f"\ninput types:\n"
   (...)
    495         f"\ninputs:\n{pprint.pformat(session_run_input)}\n{model}"
    496     ) from e

File ~/Documents/GitHub/onnx/venv/lib/python3.11/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py:220, in Session.run(self, output_names, input_feed, run_options)
    218     output_names = [output.name for output in self._outputs_meta]
    219 try:
--> 220     return self._sess.run(output_names, input_feed, run_options)
    221 except C.EPFail as err:
    222     if self._enable_fallback:

InvalidArgument: [ONNXRuntimeError] : 2 : INVALID_ARGUMENT : Non-zero status code returned while running Gather node. Name:'' Status Message: indices element out of data bounds, idx=2 must be within the inclusive range [-1,0]
```

**Comment by fengyuentau** [comment]
> y = np.take(data, indices, axis=1)

> op.Gather(data, indices)

you need to set axis=1

**Comment by justinchuby** [comment]
> you need to set axis=1

Thanks, updated

**Comment by gramalingam** [comment]
I think the documentation is fine. Note that `j_{r-2}` is a free-variable that is used in the `r-1`-th index position: we need one less than `r-1` because one of the index positions is taken up by `k`. 

Specifically, the size corresponding to the range of this free-variable is the last dimension of the input: namely 4 in your example.

Not sure if I am making myself clear, since it is hard to spell it out in words.

**Comment by gramalingam** [comment]
Since it is a free-variable, we can indeed give it names like you say. But both versions have their pros and cons.

It is easier if we fix the value of r, say to 3 and q to 2, like in your example. Then, for axis = 1, we have
```
   output[a, b, c, d] = input[a, indices[b, c], d]
```
The current documentation says:
```
   output[j0, i0, i1, j1] = input[j0, indices[i0, i1], j1]
```
You want to say:
```
   output[j0, i0, i1, j2] = input[j0, indices[i0, i1], j2]
```
But `d`, `j1`, and `j2` in these three equations are just different names for the same free-variable which ranges over the last axis of the input.

Does that make sense?

**Comment by fengyuentau** [comment]
Thank you for response and explanation. I learned your concept about free-variable, but I think for most of the people (programmers) who try to understand what Gather is doing from the documentation, they would simply assume `j1` is the second index rather than the last index in the example.

**Comment by gramalingam** [comment]
I think there is definitely room for improvement in the documentation. I agree that using j2 rather than j1 may help some readers. I think what may help even more (for me personally, at least) is to first add a few specific examples for fixed values of r and q and axis, and then present the general equation for arbitrary r an q. I find it very hard to parse and understand the general equation whether we use j1 or j2.

So, for example, something like the following may help:

| input shape | indices shape | axis | output shape | output equation |
| --- | --- | --- | --- | --- |
| (P, Q) | (R, S) | 0 | (R, S, Q) | output[r, s, q] = input[ [indices[r, s], q] |
| (P, Q) | (R, S) | 1 | (P, R, S) | output[p, r, s] = input[ p, indices[r, s]] |

etc.

**Comment by fengyuentau** [comment]
Yes, this would be helpful. Some special cases should be added as well, such as 0d indices:

| input shape | indices | axis | output shape | output equation |
| - | - | - | - | - |
| (P, Q) | r | 0 | Q | output[q] = np.squeeze(input[r, q]) |
